File: python/linked-lists/reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        fast, slow, current = head, head, head
        while fast:
            previousslow = slow
            fast = fast.next
            slow = slow.next
            if fast:
                fast = fast.next

        previousslow.next = None
        tail = []

        def reversal(node):
            if not node.next:
                tail.append(node)
                return node

            nextnode = reversal(node.next)
            nextnode.next = node
            node.next = None

            return node

        if not slow:
            return
        reversal(slow)
        reverse = tail[0]
        while current and reverse:
            temp = current.next
            current.next = reverse
            reverse = reverse.next
            current.next.next = temp
            current = current.next
            current = current.next

File: c++/linked-list/reorder-list.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        int count = 0;
        ListNode* pointer = head;
        while (pointer) {
            pointer = pointer->next;
            count += 1;
        }

        pointer = head;
        if (count % 2 == 0) {
            count /= 2;
            count -= 1;
        } else {
            count /= 2;
        }
        while (count > 0) {
            count -= 1;
            pointer = pointer->next;
        }

        ListNode* first = head;
        ListNode* second = reverseList(pointer->next);

        pointer->next = nullptr;
        
        pointer = new ListNode();
        while (first || second) {
            if (first) {
                pointer->next = first;
                first = first->next;
                pointer = pointer->next;
                pointer->next = nullptr;
            }

            if (second) {
                pointer->next = second;
                second = second->next;
                pointer = pointer->next;
                pointer->next = nullptr;
            }
        }
    }

private:
    ListNode* reverseList(ListNode* head) {
        if (!head || !head->next) {
            return head;
        }

        ListNode* prev = head;
        ListNode* curr = prev->next;

        while (curr) {
            ListNode* next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        head->next = nullptr;

        return prev;
    }
};
